### DISCUSSION TITLE
SOLVED - Have vscode-jupyter do proper diffing in Git

### DISCUSSION THREAD

**Discussion by DennisL68** [microsoft/vscode-jupyter#10742] [Questions and Answers] [answered]
When trying to use git versioning with Jupyter notebooks, you end up with heugh diffs because Jupyter uses a structured format, JSON, instead of plain text. So Git will do a binary "diff" instead.

One solution is to use [GitHub - jupyter / nbdime](https://github.com/jupyter/nbdime) along with git, but it would be so much easier if nbdime was part of the vscode-jupyter distribution. Then I wouldn't have to remember to set up nbdime when installing the Jupyter extension for vscode (git would of course be a prerequisite).

As explained at the [nbdime](https://nbdime.readthedocs.io/en/latest/) Open Source project site:

---
nbdime provides “content-aware” diffing and merging of Jupyter notebooks. It understands the structure of notebook documents. Therefore, it can make intelligent decisions when diffing and merging notebooks, such as:

* eliding base64-encoded images for terminal output
* using existing diff tools for inputs and outputs
* rendering image diffs in a web view
* auto-resolving conflicts on generated values such as execution counters

nbdime yields diffs like this:

![image](https://user-images.githubusercontent.com/20666032/178014231-696b2b76-3d22-47f0-a834-ef30ca276c74.png)
---

vscode-jupyter already pulls additional exentsions on setup, such as Jupyter Keymap and Jupyter Notebook Renderers and of course .Net Interactive Notebooks...

**Comment by DonJayamanne** [comment, marked as answer]
Currently VS Code has a great diffing tool for Jupyter Notebooks, we don't just diff the JSON file, see here
https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_custom-notebook-diffing
Please try this if you haven't, and let us know what's missing or needs to be improved here.

<img src="https://code.visualstudio.com/assets/docs/datascience/jupyter/notebook-diffing.png">

**Reply by DennisL68** [reply]
As far as I can tell, this isn't a Git diff. Right?

So git will still be bloated with complete binary copies of the Jupyter notebook JSON file...

**Reply by DonJayamanne** [reply]
> this isn't a Git diff. Right?

it is a git diff.
i suggest you try this

**Reply by DennisL68** [reply]
Thanks :)

**Reply by lgonzalezsa** [reply]
When doing the diff in vscode, could be possible to highlight the differences in same way we saw it with fr example .py files? Think is a minimap with red mark pointing what changed in the code instead of scrolling the diff to find where the code is diff, in my opinion this helps to quickly find what changed and go just to the right cell to check the code, output or meta.

**Reply by DonJayamanne** [reply]
@rebornix any idea whether this is possible today?

**Reply by rebornix** [reply]
@lgonzalezsa @DonJayamanne it's on our iteration plan and I plan to implement this in Sept.

**Reply by lalitkumarj** [reply]
Hi. For some reason I don't see "Show metadata differences" and "show output differences" when I open the file from the git pane. It just opens as JSON. Any ideas on why this might happen?

**Reply by jparta** [reply]
@rebornix Are there updates on this feature?

**Reply by 2-X** [reply]
i don't see any git features in VSCode's jupyter notebook previews at all. nor do i see settings. was this removed or am i just dumb?

**Reply by caph1993** [reply]
If you go to Explorer tab, click the file you are interested in, then click on Timeline in the bottom (below the files, in the explorer pane), then you can click on previous commits and see the changes with this special jupyter diff tool.

Sadly, this diff is not implemented everywhere (which should). For instance, I use git graph extension, and if I go and click some file in history and try to diff, the diff is in JSON mode and the options do not appear. Maybe this is what you are doing and this is why you don't see the options.

## If you access from Explorer -> Timeline: Works 😊
![2024-08-22_12-35](https://github.com/user-attachments/assets/dba9543c-2a85-4197-bf35-af8ac995f823)

## If you access from git graph extension: Does not work ☹
![2024-08-22_12-37](https://github.com/user-attachments/assets/76ed948d-a89c-44d5-b64b-f9648e19ec09)
![2024-08-22_12-38](https://github.com/user-attachments/assets/ce4da50f-8f15-40bc-b444-6ba09356df7f)

**Comment by thoughtfuldata** [comment]
How to use the diffing tool as a merge tool?

**Reply by NicolaiLolansen** [reply]
It doesnt seem like the diff tool works once a merge-conflict arises. I found a related issue here: https://github.com/microsoft/vscode/issues/160941
